### PR TITLE
✨ constraints-sha

### LINKED ISSUE #23: 🐛 constraint lockfile tracking
## Summary

Lockfiles can be out of date without knowing it with the `pip-compile-restraints` feature

## Steps to Reproduce

1. You create the default lockfile
2. You create the test lockfile with the default lockfile as its constraint
3. You update the default lockfile without actually changing the requirements specified in your pyproject.toml
4. You run the test environment and it thinks it's up to date but remains pinned on the old versions

### PR DESCRIPTION
**Pull request by juftin**
## Summary

- Adds a new field to the header, `[constraints-sha]` which is a sha256 of the contents of the constraints file. If anything on the constraints lockfile is different than when the lock was first generated it should assume it's out of date. 

- Adds a new, cached, helper property: `lockfile_up_to_date`
   - There was some sporadic validation and checking of the lockfile / constraint file at different parts of the plugin implementation. This new helper method centralizes the validation and cleans up the API a bit which was necessary to implement this (IMO 🤷 )

- Doesn't edit the existing lockfile in-place. Instead moves files in and out of tempdir

- Deletes a lockfile and cleans the environment if lockfile still exists but dependencies disappear

Resolves #23

### CODE REVIEW

File: hatch_pip_compile/lock.py [hunk 1]
```diff
@@ -44,24 +45,28 @@ def process_lock(self) -> None:
...
             lockfile_text,
         )
         if self.constraints_file is not None:
+            constraint_sha = hashlib.sha256(self.constraints_file.read_bytes()).hexdigest()
             constraints_path = self.constraints_file.relative_to(self.project_root)
-            constraints_line = f"# [constraints] {constraints_path}"
-            joined_dependencies = "\n".join([constraints_line, "#", joined_dependencies])
+            constraints_lines = [
+                f"# [constraints]: {constraints_path}",
+                f"# [constraints-sha]: {constraint_sha}",
```
Comment (oprypin) [inline hatch_pip_compile/lock.py:59]: Good feature. I was just thinking, should the design of these headers converge on being "generic" like this (one per line) or perhaps something more "cute" and "integrated", e.g.

```
[constraints] requirements.txt (SHA256: 8f0f4a2dddff0e6f0499b14b933e53eccfd7678d11339b395b044b97c348e390)
```

Comment (juftin) [reply hatch_pip_compile/lock.py:59]: I like this better

```
# [constraints] requirements.txt (SHA256: 8f0f4a2dddff0e6f0499b14b933e53eccfd7678d11339b395b044b97c348e390)
```

Comment (juftin) [reply hatch_pip_compile/lock.py:59]: Just curious, how have you been staying up to date as more upgrades are pushed to the plugin? I haven't adopted it in other projects yet.

If you install the plugin like this, is there an easy way to update? Something like `hatch env run --env=internal pip install --upgrade hatch-pip-compile` would be nice

```toml
[tool.hatch.env]
requires = [
    "hatch-pip-compile"
]
```

Comment (oprypin) [reply hatch_pip_compile/lock.py:59]: 1. I actually use local checkouts of many repos and add all of them to PYTHONPATH so it kicks in instantly after any pull or change (can't recommend though, it's the most volatile approach)

2. Generally how to upgrade - I guess just run the same pip as was used to install hatch? Or pipx: `pipx inject --pip-args=-U hatch hatch-pip-compile`

3. Or you could probably add a version constraint to the `requires` in a project?

**Pull request merged**